File: .notegraph_delta.py
import json, subprocess, sys

prev_raw = subprocess.check_output(['git', 'show', 'HEAD:notegraph.json'])
prev = json.loads(prev_raw)
cur = json.load(open('notegraph.json'))

ps = prev.get('stats', {})
cs = cur.get('stats', {})
print('previous stats:', ps)
print('current  stats:', cs)

node_delta = cs.get('nodes', 0) - ps.get('nodes', 0)
edge_delta = cs.get('edges', 0) - ps.get('edges', 0)
orphan_delta = cs.get('orphans', 0) - ps.get('orphans', 0)
bundled_delta = cs.get('bundled', 0) - ps.get('bundled', 0)

print('node_delta=' + str(node_delta))
print('edge_delta=' + str(edge_delta))
print('orphan_delta=' + str(orphan_delta))
print('bundled_delta=' + str(bundled_delta))

def compute_orphans(g):
    nodes = g.get('nodes', [])
    edges = g.get('edges', [])
    ind, outd = {}, {}
    for n in nodes:
        nid = n.get('id')
        ind[nid] = 0
        outd[nid] = 0
    for e in edges:
        s = e.get('source') or e.get('from')
        t = e.get('target') or e.get('to')
        if s in outd:
            outd[s] += 1
        if t in ind:
            ind[t] += 1
    return set(nid for nid in ind if ind[nid] == 0 and outd[nid] == 0)

def compute_bundled(g):
    return set(n.get('id') for n in g.get('nodes', []) if n.get('bundled'))

po = compute_orphans(prev)
co = compute_orphans(cur)
new_orphans = sorted(co - po)
resolved_orphans = sorted(po - co)

pb = compute_bundled(prev)
cb = compute_bundled(cur)
new_bundled = sorted(cb - pb)
resolved_bundled = sorted(pb - cb)

print('new_orphans(' + str(len(new_orphans)) + '):', new_orphans[:10])
print('resolved_orphans(' + str(len(resolved_orphans)) + '):', resolved_orphans[:10])
print('new_bundled(' + str(len(new_bundled)) + '):', new_bundled[:10])
print('resolved_bundled(' + str(len(resolved_bundled)) + '):', resolved_bundled[:10])

if new_bundled:
    verdict = str(len(new_bundled)) + ' new bundled note(s): ' + new_bundled[0] + '...'
elif new_orphans:
    verdict = str(len(new_orphans)) + ' new orphan(s): ' + new_orphans[0] + '...'
elif node_delta > 0 and orphan_delta <= 0 and bundled_delta <= 0:
    verdict = '+' + str(node_delta) + ' notes wired in'
elif edge_delta > 10:
    verdict = '+' + str(edge_delta) + ' new edges'
else:
    verdict = 'graph refreshed (' + str(cs.get('nodes',0)) + 'n / ' + str(cs.get('edges',0)) + 'e / ' + str(cs.get('bundled',0)) + 'b)'

print('VERDICT=' + verdict)

File: .notegraph_delta.mjs
import { execFileSync } from 'node:child_process';
import { readFileSync, writeFileSync } from 'node:fs';

const prevRaw = execFileSync('git', ['show', 'HEAD:notegraph.json'], { maxBuffer: 200 * 1024 * 1024, encoding: 'utf8' });
const prev = JSON.parse(prevRaw);
const curr = JSON.parse(readFileSync('notegraph.json', 'utf8'));

const p = prev.stats, c = curr.stats;
const node_delta = c.nodes - p.nodes;
const edge_delta = c.edges - p.edges;
const hard_delta = (c.hard||0) - (p.hard||0);
const soft_delta = (c.soft||0) - (p.soft||0);
const orphan_delta = (c.orphans||0) - (p.orphans||0);
const bundled_delta = (c.bundled||0) - (p.bundled||0);
const atomic_delta = (c.atomic||0) - (p.atomic||0);

const prevOrphans = new Set((prev.nodes||[]).filter(n => (n.inDegree||0)===0 && (n.outDegree||0)===0).map(n => n.id));
const currOrphans = new Set((curr.nodes||[]).filter(n => (n.inDegree||0)===0 && (n.outDegree||0)===0).map(n => n.id));
const new_orphans = [...currOrphans].filter(x => !prevOrphans.has(x));
const resolved_orphans = [...prevOrphans].filter(x => !currOrphans.has(x));

const prevBundled = new Set((prev.nodes||[]).filter(n => n.bundled).map(n => n.id));
const currBundled = new Set((curr.nodes||[]).filter(n => n.bundled).map(n => n.id));
const new_bundled = [...currBundled].filter(x => !prevBundled.has(x));
const resolved_bundled = [...prevBundled].filter(x => !currBundled.has(x));

let verdict;
if (new_bundled.length > 0) verdict = `${new_bundled.length} new bundled note(s): ${new_bundled[0]}…`;
else if (new_orphans.length > 0) verdict = `${new_orphans.length} new orphan(s): ${new_orphans[0]}…`;
else if (node_delta > 0 && orphan_delta <= 0 && bundled_delta <= 0) verdict = `+${node_delta} notes wired in`;
else if (edge_delta > 10) verdict = `+${edge_delta} new edges`;
else verdict = `graph refreshed (${c.nodes}n / ${c.edges}e / ${c.bundled}b)`;

const out = {
  prev_stats: p,
  curr_stats: c,
  deltas: { node_delta, edge_delta, hard_delta, soft_delta, orphan_delta, bundled_delta, atomic_delta },
  new_orphans, resolved_orphans, new_bundled, resolved_bundled,
  verdict
};
writeFileSync('.notegraph_delta.json', JSON.stringify(out, null, 2));
console.log(JSON.stringify(out, null, 2));
